File: sdks/typescript/src/retry.ts
export interface RetryPolicyOptions {
  initialDelayMs?: number;
  maxDelayMs?: number;
  multiplier?: number;
}

export class RetryPolicy {
  private initialDelayMs: number;
  private maxDelayMs: number;
  private multiplier: number;
  private attempt = 0;

  constructor(options: RetryPolicyOptions = {}) {
    this.initialDelayMs = options.initialDelayMs ?? 100;
    this.maxDelayMs = options.maxDelayMs ?? 30_000;
    this.multiplier = options.multiplier ?? 2.0;
  }

  nextDelay(): number {
    const delayMs = this.initialDelayMs * Math.pow(this.multiplier, this.attempt);
    const capped = Math.min(delayMs, this.maxDelayMs);
    const jitter = capped * 0.1 * Math.random();
    this.attempt++;
    return Math.floor(capped + jitter);
  }

  reset(): void {
    this.attempt = 0;
  }
}

File: sdks/go/retry.go
package valka

import (
	"math"
	"math/rand"
	"time"
)

// RetryPolicy provides exponential backoff with jitter for reconnection.
type RetryPolicy struct {
	InitialDelay time.Duration
	MaxDelay     time.Duration
	Multiplier   float64
	attempt      int
}

// DefaultRetryPolicy returns a retry policy with sensible defaults.
func DefaultRetryPolicy() *RetryPolicy {
	return &RetryPolicy{
		InitialDelay: 100 * time.Millisecond,
		MaxDelay:     30 * time.Second,
		Multiplier:   2.0,
	}
}

// NextDelay calculates the next backoff delay and increments the attempt counter.
func (r *RetryPolicy) NextDelay() time.Duration {
	delay := float64(r.InitialDelay) * math.Pow(r.Multiplier, float64(r.attempt))
	capped := math.Min(delay, float64(r.MaxDelay))
	jitter := capped * 0.1 * rand.Float64()
	r.attempt++
	return time.Duration(capped + jitter)
}

// Reset resets the attempt counter (call on successful connection).
func (r *RetryPolicy) Reset() {
	r.attempt = 0
}
